Backward compatibility of raw-converted dataset with `xr.DataTree` [all tests ci]

oftfrfbf opened the pull request:
This is for issue #1420 to reconcile backwards compatability.

Test files were added for legacy data from v0.8.4 up to v0.9.2 (technically up to version "0.9.1.dev30+gb328bf91.d20241223" since there is no tagged "0.9.2" yet). The test data was derived from an 'ek60' raw file and '.nc' and '.zarr' echodata backups were created using python 3.10.12.

I created a Pooch registry and wired the test files into the test path — if that is a problem I can rework the code and we can add the files to the google drive. In the interim that registry will eventually need to be pointed to test files living somewhere in echopype release assets and changes will need to be made to the conftest.py _base_url_ and _version_ values.

(Also sorry for some of the erroneous edits that got added. I am not sure why my pycharm environment was making so many changes, must be some problem with the way I configured the linter.)

leewujung commented:
I changed the title and closed/reopened this PR to trigger all tests.

leewujung reviewed (changes requested):
@oftfrfbf : Could you uncomment all the tests in `test_echodata.py` to make sure they run through? These were commented out in #1433 anticipating follow up from this PR. Thanks!

oftfrfbf commented:
> @oftfrfbf : Could you uncomment all the tests in `test_echodata.py` to make sure they run through? These were commented out in #1433 anticipating follow up from this PR. Thanks!

I uncommented the class "TestEchoData" and there are a couple errors ...trying to figure out what the problems are. This looked simple to fix, a comparison of two tuples where the order was different, changed it to set comparison.
```python
def test_group_paths(self, converted_zarr):
    ed = self.create_ed(converted_zarr)
    assert set(ed.group_paths) == set(self.expected_groups)
```

Other problems aren't immediately obvious. There are 3 other failed tests, e.g.:

1 of 3: test_nbytes()
```
======================== 1 failed, 6 warnings in 0.21s =========================
FAILED                       [100%]<xarray.Dataset> Size: 206kB
Dimensions:        (time_nmea: 688)
Coordinates:
  * time_nmea      (time_nmea) datetime64[ns] 6kB 2017-06-15T19:02:14.2059996...
Data variables:
    NMEA_datagram  (time_nmea) <U73 201kB ...
Attributes:
    description:  All NMEA sensor datagrams

echopype/tests/echodata/test_echodata.py:237 (TestEchoData.test_nbytes)
4690060.0 != 4688692.0
```

and 

2 of 3: test_setattr()
```
======================== 1 failed, 6 warnings in 0.24s =========================
FAILED                      [100%]<xarray.Dataset> Size: 206kB
Dimensions:        (time_nmea: 688)
Coordinates:
  * time_nmea      (time_nmea) datetime64[ns] 6kB 2017-06-15T19:02:14.2059996...
Data variables:
    NMEA_datagram  (time_nmea) <U73 201kB ...
Attributes:
    description:  All NMEA sensor datagrams

echopype/tests/echodata/test_echodata.py:276 (TestEchoData.test_setattr)
False != True
```

and

3 of 3: test_setitem()
```
======================== 1 failed, 7 warnings in 0.23s =========================
FAILED                      [100%]<xarray.Dataset> Size: 206kB
Dimensions:        (time_nmea: 688)
Coordinates:
  * time_nmea      (time_nmea) datetime64[ns] 6kB 2017-06-15T19:02:14.2059996...
Data variables:
    NMEA_datagram  (time_nmea) <U73 201kB ...
Attributes:
    description:  All NMEA sensor datagrams

echopype/tests/echodata/test_echodata.py:304 (TestEchoData.test_setitem)
['beam_group', 'beam_newname', 'channel', 'ping_time', 'range_sample'] != ['beam_newname', 'channel', 'ping_time', 'range_sample']

Expected :['beam_newname', 'channel', 'ping_time', 'range_sample']
Actual   :['beam_group', 'beam_newname', 'channel', 'ping_time', 'range_sample']
```
...this last one looks like I could just add the missing entry?!

I'll keep looking at it.

oftfrfbf commented:
Screenshot of set_attr():
<img width="1468" alt="Screenshot 2025-02-05 at 4 11 45 PM" src="https://github.com/user-attachments/assets/99fb5eea-1e77-4ec7-a57f-007b61cc503b" />

leewujung commented:
@oftfrfbf - thanks, let me know when this is ready to review!

oftfrfbf commented:
Hi @leewujung, I will continue to look at it but it might be quicker if @ctuguinay shares what the expected values are for some of these datasets. Uncommenting that test class changes the domain of what this ticket was originally trying to accomplish and now I'm trying to troubleshoot what is happening in the echodata getters and setters.

oftfrfbf commented:
For the test_nbytes it is just a situation where there is an extra variable.

For the test_setattr the problem looks like that the "Top-level" is okay but the "Sonar/Beam_group1" shouldn't have both 'x' and 'y' DataArrays. The "Sonar/Beam_group1" should only have the 'x' DataArray.
<img width="1586" alt="Screenshot 2025-02-06 at 8 25 30 PM" src="https://github.com/user-attachments/assets/85001004-5739-46de-a4e5-70ed4b4db9e0" />

leewujung commented:
@oftfrfbf : I just worry that the extra variables were generated by the new `open_converted` routine.

@ctuguinay : Could you take a look? Info in the #1420 thread might be helpful. Thanks!

oftfrfbf commented:
@ctuguinay, as a baseline, in the main branch without any of my backwards compatability changes is the expectation that the tests could be uncommented and they would all pass?

oftfrfbf commented:
So my current take with what is happening is that the "test_setattr()" is somehow setting all DataTree node items when assigning the 'x' variable, the parent and all children end up having the 'x' value assigned (e.g. VendorSpecific has 'x', Provenance has 'x', etc.). 

Somehow need to limit the 
```
try:
    node = self.__get_node(__key)
    node.dataset = __newvalue
    return self.__get_dataset(node)
except KeyError:
    raise GroupNotFoundError(__key)
```
to just set the Top-level DataTree.

ctuguinay commented:
@oftfrfbf > @ctuguinay, as a baseline, in the main branch without any of my backwards compatability changes is the expectation that the tests could be uncommented and they would all pass?

Before the #1419 PR where I commented out `TestEchoData`, it seems that they should all pass. Here's a link to a CI test in the PR prior to #1419: https://github.com/OSOceanAcoustics/echopype/actions/runs/12551439431/job/34995864795#step:12:1470

My initial impression was that getting `TestEchoData` to work again was within the scope of this PR.

ctuguinay commented:
Here's the echodata zarr corresponding to that `nbytes` test: [ed_test_nbytes_zarr.zip](https://github.com/user-attachments/files/18710775/ed_test_nbytes_zarr.zip)

ctuguinay commented:
"variables are inherited to descendent nodes" then things are acting in accordance with spec and the test is wrong?

Yeah I think you're right about the test being wrong. Inheritance should happen, but the Echodata top-level shouldn't actually have any data variables in it, as indicated here: https://echopype.readthedocs.io/en/stable/data-format-raw.html. We didn't have a problem with this test prior to the datatree migration because parent-child node inheritance wasn't enforced, but with this change now we do. So to correct this, I think top-level just shouldn't contain any data variables in it, only attributes.

@leewujung Correct me if I'm wrong about that.

ctuguinay commented:
> "variables are inherited to descendent nodes" then things are acting in accordance with spec and the test is wrong?

@oftfrfbf Also yeah, I think you're correct about this. Now that inheritance between child and parent are enforced, a lot of new stuff is introduced into Sonar/Beam_group1 from a variety of places.

![image](https://github.com/user-attachments/assets/1ba7f334-ea3d-48ad-b75c-e9bf763f8d2a)

And looking at it more, I'm not quite sure what it's trying to achieve in terms of testing setting of attributes. I am now convinced that this specific testing of attributes is probably outside of the scope of this PR.

@leewujung Any thoughts on this?

ctuguinay commented:
Also for the `nbytes` problem, since there's a lot of stuff now being inherited, I don't think it's possible to get the `nbytes` to match with what was previously there. There's both coordinate and data variable inheritances that weren't there in `Platform/NMEA` (from `Platform`) and in `Sonar/Beam_group1` from (`Beam_group1`).

ctuguinay commented:
Oh yes and here's where all the extra bytes are coming from:
![image](https://github.com/user-attachments/assets/705a23db-50eb-486b-9719-8322e84981c0)

ctuguinay commented:
> My initial impression was that getting TestEchoData to work again was within the scope of this PR and would be one of the ways to test if it was backwards compatible.

Sorry ignore this message. Took me a while to understand the problems. They actually do need fixing and I think the test nbytes can be changed to:

```
    def test_nbytes(self, converted_zarr):
        ed = self.create_ed(converted_zarr)
        assert isinstance(ed.nbytes, float)
        assert ed.nbytes == 4690060.0
```

and the test attributes is out of scope, unless there are attribute specific migration changes.

leewujung commented:
@oftfrfbf : I think there's an important word missing in your quote above:

The xarray datatree documentation says: 
> **Coordinate** variables are inherited to descendent nodes, which is only possible because variables at different levels of a hierarchical DataTree are always aligned. 

Only the coordinate variables, not all variables, otherwise this will completely disrupt the way netCDF data model is set up/used.

It is also helpful to see the entire sentence for the rationale:
> Coordinate variables are inherited to descendent nodes, which is only possible because variables at different levels of a hierarchical DataTree are always aligned.

leewujung commented:
@oftfrfbf : I pushed a merge commit to bring in all last changes in `main` properly so that only the files that actually got changed in this PR shows up under "Files changed", instead of everything since you branched out this branch. Wanted to let you know so there's no surprised if git asks you to pull first before push.

leewujung commented:
@oftfrfbf : Also, just noticed that this PR also includes adding `pooch` and related testing updates. I guess these are from addressing #1263? Do you want to separate those out from this PR, or just make this PR a double-up? (not sure if those changes are completed yet)

oftfrfbf commented:
> @oftfrfbf : Also, just noticed that this PR also includes adding pooch and related testing updates. I guess these are from addressing https://github.com/OSOceanAcoustics/echopype/issues/1263? Do you want to separate those out from this PR, or just make this PR a double-up? (not sure if those changes are completed yet)

@leewujung sorry about that, my intention was to get things put together quickly without needing to worry about adding files to the docker test image first. If it is easy to add the files to the docker image under a directory like "/test_data/ek60/legacy" or something like that I can rework the code so that it matches the way the rest of the test resources are downloaded/used.

Here is the link to the zip file:
https://github.com/oftfrfbf/echopype-test-data/releases/download/2025.2.1/legacy_data.zip

I am guessing that after the docker image is updated I can just delete the old docker images/containers and run
`python .ci_helpers/docker/setup-services.py --deploy`
again to get the updates.

leewujung commented:
> And looking at it more, I'm not quite sure what it's trying to achieve in terms of testing setting of attributes. I think there's a more targeted way to test setting of attributes that doesn't involve all the coordinates and variables. I am now thinking that this specific testing of attributes is probably outside of the scope of this PR.

@oftfrfbf @ctuguinay : I agree that there's a better way to test setting attributes. `test_setattr` seems may have been a "convenient" (lazy?) way of getting a variety of attributes to test setting them. I think it is fine to remove `test_setattr`, but we'll need a new test to target testing setting attributes. 

Looks like `x` and `y` are inherited because they are both interpreted as coordinate variables in the initialization statements `xr.Dataset({"x": [1, 2, 3, 4]})` and `xr.Dataset({"y": [4, 5, 6]})`. If they are set as data variables (so proper dimension/coordinate are provided at initialization), they won't be inherited.

With the new inheritance rule, this "convenient" way of testing attributes is problematic, since even if the `x` and `y` are properly initialized, their respective dimensions will be inherited still, creating the same problem. 

One easy way to do this is to not change things in the `Top-level` group, since it's the root for everything. If one only add  the dataset into `Sonar/Beam_group1`, it should be ok.

leewujung commented:
@oftfrfbf : For the testing: Thanks! I will add the legacy data to the docker image and trigger the action to run. I'll ping again when it's done.

I've not tried `pooch` myself - excited to learn how that works! 

I think there has been no testing file changes in the recent pushes, so chances are what you have set up before will just run without problem. 🤞

leewujung commented:
@oftfrfbf : The legacy files should be on the docker image now. I put it under `test/legacy_datatree/ek60` just so that it is clear it is for the datatree migration.

oftfrfbf commented:
@leewujung,

Okay, so I got rid of pooch, fixed test path resources w updated docker image, fixed test_nbytes, and commented out the test_setattr in my latest commits. I am unsure of what the next steps are?

leewujung commented:
Thanks @oftfrfbf! I'll review the changes and get back to you.

leewujung commented:
Oh and forgot to say: sounds good about pooch. I think once this PR is merged, we should release asap since echopype is not currently working in a fresh environment, so pretty serious problem. In this release we will need to bump a major version (so making it v0.10.0) since a data variable name changed, and being able to hande `xr.DataTree` is a major change. 

What do you think about getting `pooch` and your testing framework update in in v0.10.1?

oftfrfbf commented:
> What do you think about getting pooch and your testing framework update in in v0.10.1?

That sounds good. Just as an estimate to get pooch working for everything again would be maybe 4 or 5 hours of time needed to go through all the tests and make sure the paths are defined properly with any changes.

leewujung commented on echopype/echodata/echodata.py:
```diff
@@ -166,21 +166,36 @@ def from_file(
+        if "/Platform/NMEA" in temp_tree:
+            platform_nmea = temp_tree["/Platform/NMEA"]
+            if "time1" in platform_nmea.coords:
+                platform_nmea = platform_nmea.rename({"time1": "time_nmea"})
```
I think this should be `nmea_time` from @ctuguinay's #1419.

leewujung replied:
No worries! Thanks!

leewujung commented on echopype/test_data/README.md:
```diff
@@ -11,8 +11,6 @@ Most of these files are stored on Git LFS but the ones that aren't (due to file
 - 2019118 group2survey-D20191214-T081342.raw: Contains 6 channels but only 2 of those channels collect ping data
 - D20200528-T125932.raw: Data collected from WBT mini (instead of WBT), from @emlynjdavies
 - Green2.Survey2.FM.short.slow.-D20191004-T211557.raw: Contains 2-in-1 transducer, from @FletcherFT (reduced from 104.9 MB to 765 KB in test data updates)
-- raw4-D20220514-T172704.raw: Contains RAW4 datagram, 1 channel only, from @cornejotux
-- D20210330-T123857.raw: do not contain filter coefficients
```
Could you please add these back? It's been a longstanding problem that these kept on getting overwritten, so we try to revert the changes when we see them. We'll have a lot less problem with it later with gh assets and pooch!

oftfrfbf replied:
Changing it back. I don't understand how it was deleted.

leewujung commented on echopype/test_data/README.md:
```diff
@@ -24,7 +22,6 @@ Most of these files are stored on Git LFS but the ones that aren't (due to file
 - Winter2017-D20170115-T150122.raw: Contains a change of recording length in the middle of the file
 - 2015843-D20151023-T190636.raw: Not used in tests but contains ranges are not constant across ping times
 - SH1701_consecutive_files_w_range_change: Not used in tests. [Folder](https://drive.google.com/drive/u/1/folders/1PaDtL-xnG5EK3N3P1kGlXa5ub16Yic0f) on shared drive that contains sequential files with ranges that are not constant across ping times.
-- NBP_B050N-D20180118-T090228.raw: split-beam setup without angle data
```
Same as the above

oftfrfbf replied:
Changing it back as well.

leewujung commented on echopype/tests/echodata/test_echodata.py:
```diff
@@ -770,3 +775,22 @@ def test_echodata_chunk(chunk_dict):
                     assert chunk_sizes[-1] <= desired_chunk_size
+
+
+@pytest.mark.parametrize(
+    "legacy_datatree_filename",
+    [
+        "D20070720-T224031.raw_v0.8.4_echodata.zarr",
+        "D20070720-T224031.raw_v0.8.4_echodata.nc",
+        "D20070720-T224031.raw_v0.9.0_echodata.zarr",
+        "D20070720-T224031.raw_v0.9.0_echodata.nc",
+        "D20070720-T224031.raw_v0.9.1_echodata.zarr",
+        "D20070720-T224031.raw_v0.9.1_echodata.nc",
+        "D20070720-T224031.raw_v0.9.2_echodata.zarr",
+        "D20070720-T224031.raw_v0.9.2_echodata.nc",
+    ],
+)
```
It's great you have these versions and both zarr and netcdf!

Could you do the same using this EK80 file?
`test_data/ek80/Summer2018--D20180905-T033113.raw`
This is one with both power and complex data. We will know if there are other non-aligned coordinate variables apart from the now renamed `Platform/NMEA/time1` there. EK80 files have a lot more variations, but this one is pretty representative.

oftfrfbf replied:
Running into some problems here unfortunately. I'll share updates in a moment.

oftfrfbf replied:
@leewujung if you could please add these files to the legacy_datatree/ek80 docker image:

https://github.com/oftfrfbf/echopype-test-data/releases/download/2025.2.1/ek80_legacy_data.zip

They are ek80 netcdf & zarr backups for versions 0.8.4, 0.9.0, 0.9.1, 0.9.2 (dev version).

leewujung reviewed (commented):
@oftfrfbf : Thanks for the changes! Most of my comments are minor. The only one is to add an EK80 file to the legacy data test. 

Let me know if that one ended up giving non-alignment errors, and if so, the problematic coordinate variables. I will need to ponder about the changes if there are many.

oftfrfbf commented:
Focused on test for open_converted of ek80 with v0.9.1 "Summer2018--D20180905-T033113.raw_v0.9.1_echodata.zarr" I am getting this error:

```python
/opt/anaconda3/envs/echopype/bin/python /Users/rudy/Applications/PyCharm Professional Edition.app/Contents/plugins/python-ce/helpers/pycharm/_jb_pytest_runner.py --target test_echodata.py::test_convert_legacy_versions_ek80 
Testing started at 8:29 PM ...
Launching pytest with arguments test_echodata.py::test_convert_legacy_versions_ek80 --no-header --no-summary -q in /Users/user/Documents/github/echopype/echopype/tests/echodata

============================= test session starts ==============================
collecting ... collected 1 item

test_echodata.py::test_convert_legacy_versions_ek80[Summer2018--D20180905-T033113.raw_v0.9.1_echodata.zarr] 

======================== 1 failed, 6 warnings in 0.34s =========================
FAILED [100%]
echopype/tests/echodata/test_echodata.py:798 (test_convert_legacy_versions_ek80[Summer2018--D20180905-T033113.raw_v0.9.1_echodata.zarr])
path = '/Sonar/Beam_group1'
node_ds = <xarray.Dataset> Size: 10MB
Dimensions:                        (channel: 3, ping_time: 8,
                            ...  (channel, ping_time) <U3 288B ...
Attributes:
    beam_mode:              vertical
    conversion_equation_t:  type_3
parent_ds = <xarray.DatasetView> Size: 2kB
Dimensions:                    (beam_group: 2, channel: 5)
Coordinates:
  * beam_group ...r:     
    sonar_software_name:     EK80
    sonar_software_version:  2.0.1.0
    sonar_type:              echosounder
children = Frozen({})

    def check_alignment(
        path: str,
        node_ds: Dataset,
        parent_ds: Dataset | None,
        children: Mapping[str, DataTree],
    ) -> None:
        if parent_ds is not None:
            try:
>               align(node_ds, parent_ds, join="exact", copy=False)

/opt/anaconda3/envs/echopype/lib/python3.10/site-packages/xarray/core/datatree.py:149: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/opt/anaconda3/envs/echopype/lib/python3.10/site-packages/xarray/core/alignment.py:883: in align
    aligner.align()
/opt/anaconda3/envs/echopype/lib/python3.10/site-packages/xarray/core/alignment.py:575: in align
    self.align_indexes()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <xarray.core.alignment.Aligner object at 0x153503400>

    def align_indexes(self) -> None:
        """Compute all aligned indexes and their corresponding coordinate variables."""
    
        aligned_indexes = {}
        aligned_index_vars = {}
        reindex = {}
        new_indexes = {}
        new_index_vars = {}
    
        for key, matching_indexes in self.all_indexes.items():
            matching_index_vars = self.all_index_vars[key]
            dims = {d for coord in matching_index_vars[0].values() for d in coord.dims}
            index_cls = key[1]
    
            if self.join == "override":
                joined_index = matching_indexes[0]
                joined_index_vars = matching_index_vars[0]
                need_reindex = False
            elif key in self.indexes:
                joined_index = self.indexes[key]
                joined_index_vars = self.index_vars[key]
                cmp_indexes = list(
                    zip(
                        [joined_index] + matching_indexes,
                        [joined_index_vars] + matching_index_vars,
                        strict=True,
                    )
                )
                need_reindex = self._need_reindex(dims, cmp_indexes)
            else:
                if len(matching_indexes) > 1:
                    need_reindex = self._need_reindex(
                        dims,
                        list(zip(matching_indexes, matching_index_vars, strict=True)),
                    )
                else:
                    need_reindex = False
                if need_reindex:
                    if self.join == "exact":
>                       raise ValueError(
                            "cannot align objects with join='exact' where "
                            "index/labels/sizes are not equal along "
                            "these coordinates (dimensions): "
                            + ", ".join(f"{name!r} {dims!r}" for name, dims in key[0])
                        )
E                       ValueError: cannot align objects with join='exact' where index/labels/sizes are not equal along these coordinates (dimensions): 'channel' ('channel',)

/opt/anaconda3/envs/echopype/lib/python3.10/site-packages/xarray/core/alignment.py:422: ValueError

The above exception was the direct cause of the following exception:

legacy_datatree = PosixPath('/Users/rudy/Documents/github/echopype/echopype/test_data/legacy_datatree')
legacy_datatree_filename = 'Summer2018--D20180905-T033113.raw_v0.9.1_echodata.zarr'

    @pytest.mark.parametrize(
        "legacy_datatree_filename",
        [
            # "Summer2018--D20180905-T033113.raw_v0.8.4_echodata.nc",
            # "Summer2018--D20180905-T033113.raw_v0.8.4_echodata.zarr",
            # "Summer2018--D20180905-T033113.raw_v0.9.0_echodata.nc",
            # "Summer2018--D20180905-T033113.raw_v0.9.0_echodata.zarr",
            # "Summer2018--D20180905-T033113.raw_v0.9.1_echodata.nc",
            "Summer2018--D20180905-T033113.raw_v0.9.1_echodata.zarr",
            # "Summer2018--D20180905-T033113.raw_v0.9.2_echodata.nc",
            # "Summer2018--D20180905-T033113.raw_v0.9.2_echodata.zarr",
        ],
    )
    def test_convert_legacy_versions_ek80(legacy_datatree, legacy_datatree_filename):
        ek80_raw_path = str(legacy_datatree.joinpath("ek80", legacy_datatree_filename))
>       ed = open_converted(converted_raw_path=ek80_raw_path)

test_echodata.py:814: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../echodata/api.py:32: in open_converted
    return EchoData.from_file(
../../echodata/echodata.py:180: in from_file
    tree = xr.DataTree.from_dict(temp_tree)
/opt/anaconda3/envs/echopype/lib/python3.10/site-packages/xarray/core/datatree.py:1198: in from_dict
    obj._set_item(
/opt/anaconda3/envs/echopype/lib/python3.10/site-packages/xarray/core/treenode.py:652: in _set_item
    current_node._set(name, item)
/opt/anaconda3/envs/echopype/lib/python3.10/site-packages/xarray/core/datatree.py:944: in _set
    new_node._set_parent(new_parent=self, child_name=key)
/opt/anaconda3/envs/echopype/lib/python3.10/site-packages/xarray/core/treenode.py:115: in _set_parent
    self._attach(new_parent, child_name)
/opt/anaconda3/envs/echopype/lib/python3.10/site-packages/xarray/core/treenode.py:152: in _attach
    self._pre_attach(parent, child_name)
/opt/anaconda3/envs/echopype/lib/python3.10/site-packages/xarray/core/datatree.py:528: in _pre_attach
    check_alignment(path, node_ds, parent_ds, self.children)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

path = '/Sonar/Beam_group1'
node_ds = <xarray.Dataset> Size: 10MB
Dimensions:                        (channel: 3, ping_time: 8,
                            ...  (channel, ping_time) <U3 288B ...
Attributes:
    beam_mode:              vertical
    conversion_equation_t:  type_3
parent_ds = <xarray.DatasetView> Size: 2kB
Dimensions:                    (beam_group: 2, channel: 5)
Coordinates:
  * beam_group ...r:     
    sonar_software_name:     EK80
    sonar_software_version:  2.0.1.0
    sonar_type:              echosounder
children = Frozen({})

    def check_alignment(
        path: str,
        node_ds: Dataset,
        parent_ds: Dataset | None,
        children: Mapping[str, DataTree],
    ) -> None:
        if parent_ds is not None:
            try:
                align(node_ds, parent_ds, join="exact", copy=False)
            except ValueError as e:
                node_repr = _indented(_without_header(repr(node_ds)))
                parent_repr = _indented(dims_and_coords_repr(parent_ds))
>               raise ValueError(
                    f"group {path!r} is not aligned with its parents:\n"
                    f"Group:\n{node_repr}\nFrom parents:\n{parent_repr}"
                ) from e
E               ValueError: group '/Sonar/Beam_group1' is not aligned with its parents:
E               Group:
E                   Dimensions:                        (channel: 3, ping_time: 8,
E                                                       range_sample: 6754, beam: 4)
E                   Coordinates:
E                     * beam                           (beam) <U21 336B '1' '2' '3' '4'
E                     * channel                        (channel) <U25 300B 'WBT 545612-15 ES200-7...
E                     * ping_time                      (ping_time) datetime64[ns] 64B 2018-09-05T...
E                     * range_sample                   (range_sample) int64 54kB 0 1 2 ... 6752 6753
E                   Data variables: (12/26)
E                       angle_offset_alongship         (channel) float64 24B ...
E                       angle_offset_athwartship       (channel) float64 24B ...
E                       angle_sensitivity_alongship    (channel) float64 24B ...
E                       angle_sensitivity_athwartship  (channel) float64 24B ...
E                       backscatter_i                  (channel, ping_time, range_sample, beam) float64 5MB ...
E                       backscatter_r                  (channel, ping_time, range_sample, beam) float64 5MB ...
E                       ...                             ...
E                       transceiver_software_version   (channel) <U4 48B ...
E                       transmit_duration_nominal      (channel, ping_time) float32 96B ...
E                       transmit_frequency_start       (channel, ping_time) float64 192B ...
E                       transmit_frequency_stop        (channel, ping_time) float64 192B ...
E                       transmit_power                 (channel, ping_time) float64 192B ...
E                       transmit_type                  (channel, ping_time) <U3 288B ...
E                   Attributes:
E                       beam_mode:              vertical
E                       conversion_equation_t:  type_3
E               From parents:
E                   Dimensions:     (beam_group: 2, channel: 5)
E                   Coordinates:
E                     * beam_group  (beam_group) <U11 88B 'Beam_group1' 'Beam_group2'
E                     * channel     (channel) <U25 500B 'WBT 545612-15 ES200-7C_ES' ... 'WBT 7438...

/opt/anaconda3/envs/echopype/lib/python3.10/site-packages/xarray/core/datatree.py:153: ValueError

Process finished with exit code 1
```

leewujung commented:
@oftfrfbf : Ah ok, this is because the `channel` coordinate/dimension in the parent (`Sonar`) group has 5 channels, and one child group has 3 channels and the other has 2 channels (i.e., some channels collect power/angle data, some channels collect complex data, see [here](https://echopype.readthedocs.io/en/stable/data-format-raw.html#scenario-3)). I think the easiest is to change `channel` in the `Sonar` group to be `channel_all`, so that there is no conflicts here.

The ek80 legacy dataset is now in the docker image!

oftfrfbf commented:
> The ek80 legacy dataset is now in the docker image!

@leewujung thank you for adding that!

> I think the easiest is to change channel in the Sonar group to be channel_all, so that there is no conflicts here.

Is this something I should follow up on? If so, where would it need to happen? I would assume in the "from_file" method to fix legacy versions I would need to do another rename operation, but it would also need to happen in the echodata object as well elsewhere?

leewujung commented:
> > I think the easiest is to change channel in the Sonar group to be channel_all, so that there is no conflicts here.
> 
> Is this something I should follow up on? If so, where would it need to happen? I would assume in the "from_file" method to fix legacy versions I would need to do another rename operation, but it would also need to happen in the echodata object as well elsewhere?
> 
> (I will be offline from Feb 15th to the 23rd p.s.)

Ah you're right, I somehow missed the part about the initial setting in the Echodata object.

Could you please make the change to fix legacy versions? and I'll pick up from there to update the part to set this variable when assembling the EchoData.

Also, thinking about the logic again, we can avoid opening the data twice by only reading the `Sonar` group using `xr.open_dataset(FILENAME/Sonar, ...)`, and do `open_datatree` if it is the new data (containing `channel_all`) or do `open_groups` and the associated ops you have if it is the legacy data (containing `channel`). Do you think this would work?

oftfrfbf commented:
> Could you please make the change to fix legacy versions? and I'll pick up from there to update the part to set this variable when assembling the EchoData.

Made some changes to accommodate this. I noticed that the children were also changed from 'channel' to 'channel_all' when I renamed the variable, I am not entirely sure if that is what you were looking to happen.

I was having some sort of problem running the tests with versions 0.9.2 of the test files, I am not sure why that is so if you could please troubleshoot and resolve. Maybe it has to do with the test files being dev snapshot conversions that are out of date now (might have to recreate them with the latest dev 0.9.2 version of echopype).

> Also, thinking about the logic again, we can avoid opening the data twice by only reading the Sonar group using xr.open_dataset(FILENAME/Sonar, ...), and do open_datatree if it is the new data (containing channel_all) or do open_groups and the associated ops you have if it is the legacy data (containing channel). Do you think this would work?

This makes sense to me. It would be nice to test it out to see what the performance differences are, I am out of time before leaving town though so this will have to wait for another day. I will offline until the 22nd. Also, things are getting very very grim with respect to university institute funding and my federal institution so I will have to see if we still exist a week from now. Good luck with continued development.

leewujung commented:
> Maybe it has to do with the test files being dev snapshot conversions that are out of date now (might have to recreate them with the latest dev 0.9.2 version of echopype)

By "the latest dev 0.9.2 version of echopype" do you mean what's in `main`?


> I noticed that the children were also changed from 'channel' to 'channel_all' when I renamed the variable, I am not entirely sure if that is what you were looking to happen.

I think it may need to be a roundabout to make xarray recognizing the channels are not supposed to be inherited, something like:
- changing `channel` in the child groups to `channel_new`
- changing `channel` in the parent group to `channel_all`
- changing `channel_new` back to `channel`


> Things are also getting very very grim with respect to university institute funding and my federal institution so I will have to see if we still exist a week from now. Good luck with continued development.

Sigh, I hope things will go the right way... 🤞  Thanks for all of your help!

oftfrfbf commented:
'By "the latest dev 0.9.2 version of echopype" do you mean what's in main?'

Yeah, I didn't have a lot of time to figure out what the problem was - or if there even was one. Maybe I just didn't have the latest updates synced properly.

leewujung commented:
So the `channel` to `channel_all` actually was not the problem for the failed tests. The problem lies in assuming all echodata files one tries to open has `time1` in the `Platform/NMEA` group. The check should be on if `time1` (rather than `nmea_time`) exists in that group. The same goes for checking if `channel` (rather than `channel_all`) exists in `Sonar`.

I'll get this fixed using the approach I proposed above to open individual group instead of `open_datatree` or `open_groups` to avoid opening the dataset twice. Let's see if it works out!

leewujung commented:
>> By "the latest dev 0.9.2 version of echopype" do you mean what's in main?
> 
> Yeah, I didn't have a lot of time to figure out what the problem was - or if there even was one. Maybe I just didn't have the latest updates synced properly.

Ok, I think we can do without this, since there isn't a v0.9.2 release!

ctuguinay commented on echopype/tests/echodata/test_echodata.py:
```diff
@@ -192,155 +195,145 @@ def compute_range_samples(request, test_path):
+        ed["Sonar/Beam_group1"] = ed["Sonar/Beam_group1"].rename({"beam": "beam_newname"})
 
-#         assert sorted(ed['Sonar/Beam_group1'].dims.keys()) == ['beam_newname', 'channel', 'ping_time', 'range_sample']
+        assert sorted(ed["Sonar/Beam_group1"].dims.keys()) == [
```
```suggestion
        assert sorted(ed["Sonar/Beam_group1"].sizes.keys()) == [
```

ctuguinay replied:
Since dimensions will not be returning dictionaries anymore.

ctuguinay commented on echopype/echodata/echodata.py:
```diff
@@ -168,21 +169,90 @@ def from_file(
+            # Update Sonar for all sonar_model if channel exists as a coordinate
+            sonar = temp_tree["/Sonar"]
+            if "channel" in sonar.coords:
+                # TODO: do we actually want to rename the children as well?
```
Correct me if I'm wrong, but everything must be renamed to ensure that the datatree enforced parent-child inheritance holds if it's read in again right?

ctuguinay replied:
And from #1419, I  think that the "channel_all" renaming should only be done to the EK80 files and not the EK60 files.

ctuguinay replied:
Sorry I should've also highlighted line 224.

ctuguinay replied:
@leewujung Ah I see, I think I agree with you on this. Right now, the set groups for EK60 doesn't have "channel_all" in `Sonar/Beam_groupX`. I can add that to ensure that we have uniform naming across both EK60 and EK80 set groups.

leewujung replied:
Good catch - I have overlooked that missing variable... yes please add it in!

leewujung replied:
Ok @ctuguinay and I just talked and decided to keep it as is to not introduce confusion. We'll take on a review of all variables at another time.

ctuguinay commented:
@leewujung I just had a few comments on the legacy checker/fixing code. Let me know if they're correct and I can go in and modify if needed.

leewujung commented:
@ctuguinay : Thanks for reviewing this, and making the change for `.dims` to `.sizes`. I responded to you comments above. Let me know what you think.

leewujung commented:
Ok, we've resolved all points here now and I'll go ahead to merge. Thanks @oftfrfbf and @ctuguinay for your work on this!
